File: packages/eyes-universal/src/spec-driver.js
const {TypeUtils} = require('@applitools/eyes-sdk-core')

function makeSpecDriver(ws) {
  // #region UTILITY
  function isDriver(driver) {
    return true
  }
  function isElement(element) {
    return Boolean(element && element['element-ref-id'])
  }
  function isSelector(selector) {
    return TypeUtils.isString(selector) || TypeUtils.has(selector, ['type', 'selector'])
  }
  function extractSelector(element) {
    return element.selector
  }
  function isStaleElementError(error) {
    // error
  }
  // #endregion

  // #region COMMANDS
  async function isEqualElements(context, element1, element2) {
    return ws.request('Driver.isEqualElements', {context, element1, element2})
  }
  async function executeScript(context, script, ...args) {
    return ws.request('Driver.executeScript', {context, script: script.toString(), args})
  }
  async function mainContext(context) {
    return ws.request('Driver.mainContext', {context})
  }
  async function parentContext(context) {
    return ws.request('Driver.parentContext', {context})
  }
  async function childContext(context, element) {
    return ws.request('Driver.childContext', {context, element})
  }
  async function findElement(context, selector) {
    return ws.request('Driver.findElement', {context, selector})
  }
  async function findElements(context, selector) {
    return ws.request('Driver.findElements', {context, selector})
  }
  async function getWindowRect(driver) {
    return ws.request('Driver.getWindowRect', {driver})
  }
  async function setWindowRect(driver, rect) {
    return ws.request('Driver.setWindowRect', {driver, rect})
  }
  async function getViewportSize(driver) {
    return ws.request('Driver.getViewportSize', {driver})
  }
  async function setViewportSize(driver, size) {
    return ws.request('Driver.setViewportSize', {driver, size})
  }
  async function getOrientation(driver) {
    return ws.request('Driver.getOrientation', {driver})
  }
  async function getTitle(driver) {
    return ws.request('Driver.getTitle', {driver})
  }
  async function getUrl(driver) {
    return ws.request('Driver.getUrl', {driver})
  }
  async function getDriverInfo(driver) {
    return ws.request('Driver.getDriverInfo', {driver})
  }
  async function takeScreenshot(driver) {
    const buffer = await ws.request('Driver.takeScreenshot', {driver})
    return Buffer.from(buffer.data)
  }
  // #endregion

  return {
    isDriver,
    isElement,
    isSelector,
    extractSelector,
    isStaleElementError,
    isEqualElements,
    executeScript,
    mainContext,
    parentContext,
    childContext,
    findElement,
    findElements,
    getWindowRect,
    setWindowRect,
    getViewportSize,
    setViewportSize,
    getOrientation,
    getTitle,
    getUrl,
    getDriverInfo,
    takeScreenshot,
  }
}

module.exports = makeSpecDriver

File: rb/eyes-selenium/lib/applitools/spec-driver.rb
module Applitools
  module SpecDriver
    extend self

    def isEqualElements(driver, element1, element2)
      element1.hash == element2.hash
    end

    def executeScript(driver, script, *args)
      _script = script.start_with?('return') ? script : "return (#{script}).apply(null, arguments)"
      driver.execute_script(_script, *args)
    end

    def mainContext(driver)
      driver.switch_to.default_content
    end

    def parentContext(driver)
      driver.switch_to.parent_frame
    end

    def childContext(driver, element)
      driver.switch_to.frame(element)
    end

    def findElement(driver, selector)
      driver.find_element(transformSelector(selector))
    end

    def findElements(driver, selector)
      driver.find_elements(transformSelector(selector))
    end

    def getWindowRect(driver)
      driver.manage.window.rect
    end

    def setWindowRect(driver, rect)
      if rect[:width] && rect[:height]
        driver.manage.window.resize_to(rect[:width], rect[:height])
      end
      if rect[:x] && rect[:y]
        driver.manage.window.move_to(rect[:x], rect[:y])
      end
    end

    def getTitle(driver)
      driver.title
    end

    def getUrl(driver)
      driver.current_url
    end

    def getDriverInfo(driver)
      caps = driver.capabilities
      {
        :platformName => caps.platform,
        :isMobile => caps.platform.include?(('android' or 'ios'))
      }
    end

    def commands
      self.instance_methods.map {|method_name| method_name.to_s}
    end

    private

      def transformSelector(selector)
        return {css: selector}
      end

  end
end
